Feature/post - Posts CRUD + Filtering

elliotyim opened the pull request:
Posting의 기본적인 CRUD 기능과 필터링 기능을 추가하였습니다.

현재 이 PR에 너무 많은 작업이 들어가 있어서
Swagger 주석 작업과 페이징은 이 PR이 반영되고 나면 따로 PR을 만들어서 작업하도록 하겠습니다.

elliotyim committed:
Modify on_delete options of foreign key fields

A few foreign key fields' lookups are rollback as it should not be null=False.
If this is not set as PROTECT or any substitute,
it will cause an error when you try to migrate.

hyeonjames commented:
음, 왠만하면 다 되었을 때 리뷰 요청해주시는게 좋겠어요, 알림이 따로 안오거든요 ㅜ

elliotyim committed:
Remove blank lines

PEP 257 shows any blank lines are required after docstring.

elliotyim committed:
Add category test

Simple test codes of the category api.

elliotyim committed:
Add category serializer

elliotyim committed:
Add routers from the post app

elliotyim committed:
Add category api

elliotyim committed:
Add two additional test cases

- Test creating the same title category that already exists
- Test getting detail of the category that does not exist.

elliotyim committed:
Fix typos

elliotyim committed:
Fix response status code properly

After deletion, status code should be 204(No Content).

elliotyim committed:
Fix category delete() & refactor redundancy

- Removed ParseError in delete().
- Removed parameters `title` as `title_data`s are doing their job.

elliotyim committed:
Add tag tests

elliotyim committed:
Implement a tag serializer

elliotyim committed:
Add routers for the tag api

elliotyim committed:
Implement tag api

elliotyim committed:
Relocate fixtures in conftest.py

This seems the default way pytest loads fixtures.

elliotyim committed:
Update user model from basic one to customised one

elliotyim committed:
Apply model mommy to make fixtures

elliotyim committed:
Change url patterns as rest api standards

elliotyim committed:
Fix the missed types

elliotyim committed:
Fix import order for isort

elliotyim committed:
Remove put method of Tag model

Tags should not be modified for those who are using it but did not create it.

elliotyim committed:
Remove `/` at the end of the urls

One of the rule of Rest API is that: "Do not contain `/` at the end"
Moreover, containing `/` at the end may cause an unpredictable error.
(e.g. DELETE /posts/3 -> GET /posts/3/)

elliotyim committed:
Add fixtues for posting

elliotyim committed:
Add simple test codes for posting

elliotyim committed:
Add Serializers

There are two serializers for post model, one for posting and another one is for GET
requests.

elliotyim committed:
Modify date field type of the post model

Datetime is redundant for the date field.

elliotyim committed:
Add post api

elliotyim committed:
Solve isort problem

elliotyim committed:
Add tag update test

The tag api returns a 405 error when the `update` method is called as tags should not be modified.

elliotyim committed:
Modify category api in order to include status code 201

Category api should include status code 201 instead of 200 when a category is modified.

elliotyim committed:
Modify post api

Change log:

Post API
- Moved validate() into APIUtils class to remove redundancy.
- Make post api check if there is category number in request data.
- Modify post api in order to include status code 201.

API Utils
- Added validate() into APIUtils class.

elliotyim committed:
Add comment tests

elliotyim committed:
Add comment serializers

elliotyim committed:
Add comment views into the router

elliotyim committed:
Add fixtures into the conftest module for comment tests

elliotyim committed:
Implement comment api

elliotyim committed:
Modify model's fields settings

elliotyim committed:
Fix mypy errors

elliotyim committed:
Add category, tag filtering tests

elliotyim committed:
Add post preset data and refactor fixture codes

elliotyim committed:
Implement category, tag filtering

If a querystring is added when sending a HTTP request,
the api will filter result data with input parameters.

elliotyim committed:
Refactor test codes

Change Log:
- Unified all the `id` fixtues into `dummy_id`.
- Used _quantity lookup to make multiple mocks.
- Removed the blank line on post.py

elliotyim committed:
Add a few more filtering test codes

Change log:
- Added start date & end date filtering test codes
- Added time of day filtering test codes
- Added location filtering test codes
- Polish codes

elliotyim committed:
Implement a few more filtering features

Change log:
- Added start date & end date filtering feature.
- Added time of day filtering feature.
- Added location filtering feature.

elliotyim committed:
Fix type hinting and add dependencies

Change Log:
- Added dependency that would be added from feature/settings branch later
 : djangorestframework-stubs, model-bakery
- Removed all `# type: ignore`s
- Modified all `Any` types if they should not be `Any` type.

elliotyim committed:
Modify model mommy

Model mommy have been replaced by model bakery that is almost the same library.

elliotyim committed:
Added parent comment into vulneralble fields

Change Log:
- Added a test that tests if a request trys to change a parent comment of a comment.
- Added parent comment into vulnerable fields.

elliotyim committed:
Added few more vulnerable fields.

elliotyim committed:
Polish api doc string

elliotyim committed:
Add a default value into the paret comment field of the comment model

elliotyim committed:
Change all the snake case parameter name & result data into camel case

elliotyim committed:
Remove unnecessary serializer

elliotyim committed:
Refactor redundant codes

elliotyim committed:
Rename the pakage name of post_test

elliotyim committed:
Add constants of time of day fixture to clarify its meaning

elliotyim committed:
Polish codes

Change Log:
- Post filtering test codes are separated from Post CRUD test codes
- Refactoring

hyeonjames commented on dev_up/urls.py:
```diff
@@ -33,4 +33,5 @@
     path('swagger/', schema_view.with_ui('swagger', cache_timeout=0), name='schema-swagger-ui'),
     path('redoc/', schema_view.with_ui('redoc', cache_timeout=0), name='schema-redoc'),
     path('user/', include('user.routers')),
+    path('', include('post.routers')),
```
저번에 말했던 `/posts/post` 때문에 이렇게 한건가요?
차라리 posts/urls.py에서 post쪽만 지우는게 어떤가요?

elliotyim replied:
그런 방법이!?
역시 천재적이시군요...
왜 난 그 생각을 못했지? ㄷㄷ

elliotyim committed:
Refactor serializer

- Removed unnecessary codes.

hyeonjames commented on post/models.py:
```diff
@@ -46,13 +44,12 @@ class Comment(models.Model):
-    post = models.ForeignKey('Post', on_delete=models.SET_NULL, null=True)
-    parent_comment = models.ForeignKey('self', on_delete=models.PROTECT, null=True)
+    post = models.ForeignKey('Post', on_delete=models.CASCADE)
+    parent_comment = models.ForeignKey('self', on_delete=models.SET_NULL, null=True, default=None)
```
parent_comment를 SET_NULL으로 바꾸자는 얘기가 된적이 없어요 '-'

elliotyim replied:
저게 제가 바꾸다가 잘못해서 PROTECT가 되어있는줄 알고 SET_NULL로 했었는데 그게 아니라 원래 쟤만 PROTECT가 맞았던거군요? 다시 바꾸겠습니다.

hyeonjames commented on post/api/category.py:
```diff
@@ -0,0 +1,90 @@
+        """
+        try:
+            return Category.objects.get(pk=category_id, is_active=True)
+        except Exception:
```
`get` 함수가 무조건 없을때만 Exception이 일어난다고 보장 못합니다.
`DoesNotExist` exception 만 잡아서 NotFound 해주세요 (나머지 APIView들도)

elliotyim replied:
넵 수정하겠습니다

hyeonjames commented on post/api/category.py:
```diff
@@ -0,0 +1,90 @@
+
+        return Response(serializer.data)
+
+    def get(self, request: Request, **url_resources: Optional[int]) -> Response:
```
`category_id: Optional[int]` 로 하지 않은 이유가 따로 있을까요?
그게 아니라면 그냥 `category_id : Optional[int]` 형식이 좋을 것 같아요.

elliotyim replied:
이 부분은 제가 저 파라미터에 대한 이해도가 떨어져서 그랬습니다.
다른 인자가 올 수도 있을거라 생각해서 그랬어요.
바로 수정하겠습니다.

elliotyim replied:
방금 수정하다보니 알게 된건데, url_resources를 category_id로 하게 될 경우
아래 48번째 줄에서 중복된 느낌의 코드가 써집니다.

예)
category_id = category_id.get('category_id')
또는 
category_id = category_id['category_id']

뭔가 세번 연속으로 category_id가 중복되는건 좋지 못한 것 같아요.
차라리 url_resources를 다른 이름으로 바꾸는게 낫지 않을까 싶습니다.

hyeonjames replied:
> 방금 수정하다보니 알게 된건데, url_resources를 category_id로 하게 될 경우
> 아래 48번째 줄에서 중복된 느낌의 코드가 써집니다.
> 
> 예)
> category_id = category_id.get('category_id')
> 또는
> category_id = category_id['category_id']
> 
> 뭔가 세번 연속으로 category_id가 중복되는건 좋지 못한 것 같아요.
> 차라리 url_resources를 다른 이름으로 바꾸는게 낫지 않을까 싶습니다.

`category_id` 자체가 이미 `Optional[int]`인데 `get`을 쓸 이유가 없지 않나요

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,200 @@
+        post3 = api_client.post('/posts', post3_data).data
+        assert post1 != post2 and post1 != post3 and post2 != post3
+
+        (category_id, tag_id) = (categories[0].id, tags[1].id)
```
() 없어두 되요

elliotyim replied:
넵

hyeonjames commented on tests/post/conftest.py:
```diff
@@ -0,0 +1,139 @@
+
+
+@pytest.fixture
+def dummy_id() -> int:
```
이 fixture 같은 경우는, 오히려 복잡도가 증가해서 손해일 것 같아요.

elliotyim replied:
fixture보다 그냥 숫자 1을 쓰는게 낫다는거죠?

hyeonjames replied:
> fixture보다 그냥 숫자 1을 쓰는게 낫다는거죠?

넵, 일반적인 데이터들은 fixture를 쓸 필요가 없다고 생각해요. 특히나 테스트이기 때문에 디자인 적인 이슈보다 읽기 쉬운게 더 좋을 것 같아요.

hyeonjames commented on tests/post/conftest.py:
```diff
@@ -0,0 +1,139 @@
+
+@pytest.fixture
+def title_data() -> dict:
+    return {'title': 'test_title'}
```
되도록 내용이 짧은거, 혹은 두번이상 쓰이지 않는 것들은 fixture로 하지 않는게 좋을것 같아요.

elliotyim replied:
넵

elliotyim committed:
Fix post filtering bug

Change Log:
Tag filtering feature support two ways to provide `tags` parameter
(Tag parameters mean tag IDs)
1. /posts?tags=1&tags=2
2. /posts?tags=1,2 (or /posts?tags=1, 2)

It also distinguish 1&2 and 12.

elliotyim committed:
Remove redundant function calling

elliotyim committed:
Block tag parameter when post filtering such as tags=1&tags=2

elliotyim commented:
현재 post.py 모듈에 있는 PostAPI에서 get(), post(), delete() 등 HTTP 메소드만 남기고 다른 get_object(), get_option_filtered_queryset()등의 함수들은 다른 모듈로 분리하려고 하는데 어떨까요? (다른 모듈들도 그렇게 나누려고 합니다.)

elliotyim committed:
Modify constant's name

hyeonjames commented on post/api/category.py:
```diff
@@ -0,0 +1,90 @@
+
+
+class CategoryAPI(APIView):
+    def get_object(self, category_id: int) -> Category:
```
`get_object`라는 이름이 아예 말이 안되는 건 아닌데, `get_one` 같은 것이 오히려 이해하기가 쉬울 것 같아요 
아 그리고, 내부에서만 쓰이는 함수들은 앞에 `_`를 붙여서 private표시를 해줍시다

elliotyim replied:
private 깜빡했네요. 근데 이제 이걸 APIView에서 다른 모듈로 분리할 때는 붙이면 안될 것 같아서 만약 상속해서 쓴다면 protected로 해두겠습니다.
이름도 get_one으로 바꿀게요

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,141 @@
+            time_of_day = request.GET['timeOfDay']
+            queryset = queryset.filter(time_of_day=time_of_day)
+
+        if 'location' in request.GET:
```
반복되는 코드가 너무 많아요. for문을 이용해서 확장성 높게 만들 수 있을 것 같네요

elliotyim replied:
흠...한번 해보겠습니다.

hyeonjames replied:
tags만 빼고 이렇게 하는 방법이 있죠.
```python
filters = dict(
     date__gte=request.GET.get('startDate'),
     date__lte=request.GET.get('endDate')
     .....
)
queryset = queryset.filter(**{k: v for k,v in filters.items() if v is not None })
```

hyeonjames replied:
~~이건 확실하지 않은데, `Post.objects.all()`을 맨 처음으로 써주는 거에대한 퍼포먼스 이슈에 대한건 생각해보셨나요?
만약 동기적으로 다 가져오는 거라면 `all()`을 쓰면 안될거 같거든요..~~

역시`lazy 방식인가 보네요 그럼 상관없으니 그냥 무시해주세여

hyeonjames replied:
> 이건 확실하지 않은데, `Post.objects.all()`을 맨 처음으로 써주는 거에대한 퍼포먼스 이슈에 대한건 생각해보셨나요?
> 만약 동기적으로 다 가져오는 거라면 `all()`을 쓰면 안될거 같거든요..

역시 `lazy` 방식이군요. 그냥 무시해주세여

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,141 @@
+        except Exception:
+            raise NotFound
+
+    def get_option_filtered_queryset(self, request: Request) -> 'QuerySet[Post]':
```
함수 이름을 좀 더 짧게 했으면 좋을 것 같아요

elliotyim replied:
넵 고민해보겠습니다.

hyeonjames commented on post/api/category.py:
```diff
@@ -0,0 +1,90 @@
+        The specific category ID must be required by uri resources.
+        """
+        category = self.get_object(category_id)
+        serializer = CategorySerializer(category, request.data)
```
업데이트 혹은 INSERT 작업에서, `request.data` 전체를 던지는 건 좋은 것 같지 않아요.
`request.data` 쪽에서 우리가 원하는 컬럼만을 가져올 수 있는 형태로 짜는게 좋아요. 나머지는 무시하구요.

hyeonjames replied:
밑에 `APIUtils.validate` 방식도 마찬가지에요. 물론 변경되면 안되는 컬럼들은 막아준다고 생각하시겠지만, 보안적인 이유로 우리가 받을 컬럼만 지정해주는 게 좋습니다.

elliotyim replied:
저게 사실 ListSerializer랑 그냥 Serializer랑 나눠서 사용하는 이유가 그건데,
Serializer에서 우리가 필요한 컬럼만을 지정하게 할 수 있습니다.
만약 다른 컬럼이 request.data에 들어갔다거나 하면 serializer.is_valid()에서 오류나서 에러 띄우게 되거든요.
여기서 필요한 컬럼들을 분리하는 작업을 코딩하는 것보다 serializer에서 지정해주고 민감한 정보는 update시에는 막도록 하게 하는게 더 깔끔하게 코딩할 수 있을거라 생각했습니다.

hyeonjames replied:
> 저게 사실 ListSerializer랑 그냥 Serializer랑 나눠서 사용하는 이유가 그건데,
> Serializer에서 우리가 필요한 컬럼만을 지정하게 할 수 있습니다.
> 만약 다른 컬럼이 request.data에 들어갔다거나 하면 serializer.is_valid()에서 오류나서 에러 띄우게 되거든요.
> 여기서 필요한 컬럼들을 분리하는 작업을 코딩하는 것보다 serializer에서 지정해주고 민감한 정보는 update시에는 막도록 하게 하는게 더 깔끔하게 코딩할 수 있을거라 생각했습니다.

그건 저도 알고 있어요. 근데 저는 그게 깔끔하다고 느껴지지가 않네요. 일단 `APIUtils.validate`에 종속이 된다는 점이 있고, `APIUtils.validate`에 컬럼 넣는걸 깜빡한다거나 그러면 심각해질 수 있으니까요.
차라리 업데이트 용 Serializer를 하나 더 넣거나, API 쪽에서 명시적으로 제한 시켜주는게 더 낫지 싶습니다.

hyeonjames replied:
아, 그리고 `validate` 같은 함수는 차라리 decorator 형식으로 만드는 걸 추천합니다

hyeonjames commented on tests/post/conftest.py:
```diff
@@ -0,0 +1,139 @@
+        'timeOfDay': 1,
+        'author': 1,
+        'category': 1,
+        'tags': [1, 2],
```
이거 안바꼇네요

elliotyim replied:
어떤걸 말씀하시는건지 잘모르겠습니다.

hyeonjames commented on tests/post/conftest.py:
```diff
@@ -0,0 +1,139 @@
+        **post_data,
+        'author': users[0].id,
+        'category': categories[0].id,
+        'tags': [tags[0].id]
```
마찬가지

elliotyim replied:
태그? 관련된건가요? 잘 모르겠습니다.

hyeonjames commented on tests/post/conftest.py:
```diff
@@ -0,0 +1,139 @@
+def bad_post_data() -> dict:
+    return {
+        'title': 'test_title',
+        # 'content': 'test_content',
```
안쓰는 코드는 지워주세요

elliotyim replied:
아 이건 안쓰는 코드라기보다 원래 저런 데이터가 들어가야 하는데 넣지 않았다고 시각적으로 표시해주기 위해서 주석처리를 한거긴 한데, 한눈에 알 수 있을 정도로 명확하지가 않은걸 보면 그냥 지우는게 낫겠군요.

hyeonjames commented on tests/post/test_category.py:
```diff
@@ -0,0 +1,55 @@
+        response = api_client.put(f'/categories/{dummy_id}', data=title_data)
+
+        assert response.status_code == 201
+        assert response.data.get('id') == dummy_id
```
여기에 get을 쓴 특별한 이유가 있나요? 그게 아니라면 response.data['id'] 를 쓰시는게 좋을 것 같네요.
다른 부분도 마찬가지 입니다.

elliotyim replied:
넵. 안그래도 이부분 물어볼까했었는데 전부 그렇게 바꾸겠습니다.

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,218 @@
+                                        users, categories, tags,
+                                        post_preset_data):
+
+        [post1_data, post2_data, post3_data] = post_preset_data
```
[] 없어도 되니까 그냥 지워주세요.

elliotyim replied:
넵

hyeonjames reviewed (changes requested):
내용이 너무 많아서 일단 이정도만 먼저 드려요

hyeonjames commented:
> 현재 post.py 모듈에 있는 PostAPI에서 get(), post(), delete() 등 HTTP 메소드만 남기고 다른 get_object(), get_option_filtered_queryset()등의 함수들은 다른 모듈로 분리하려고 하는데 어떨까요? (다른 모듈들도 그렇게 나누려고 합니다.)

한 모델에 종속적이게만 하지 않는다면 괜찮을 것 같아요.

elliotyim reviewed (commented):
뭐지? 왜 내가 내 코드를 리뷰 하고 있지?

elliotyim committed:
Modify uri paths

Now all uri paths of post api start `/posts/` (+1 squashed commits)

Squashed commits:

[3622145] Modify uri paths

Now all uri paths of post api start `/posts/`

elliotyim committed:
Fix uris in tests

elliotyim committed:
Modify parent comment field

Set on_delete=models.PROTECT

elliotyim committed:
Separate functions not related `HTTP Methods` from APIView classes

APIView classes have only `HTTP Methods` now and other functions are
moved into APIUtils.

elliotyim committed:
Polish test codes

This refactoring is not fully covering post test codes at the moment.

elliotyim commented:
지난 리뷰에서 아래의 세 가지를 제외하고 수정사항 반영했습니다.
(중간 저장 겸 푸시했습니다.)

1 - validate() 부분 변경
- Serializer에서 처리하도록 변경할 예정입니다.

2 - url_resources 명칭 변경
- 공부중입니다.

3 - conftest.py에 있는 post_data & post_preset_data
- 리뷰 코멘트 내용이 불분명합니다.

elliotyim committed:
Fix isort bug

elliotyim committed:
Remove the unnecessary line

hyeonjames commented:
> 3 - conftest.py에 있는 post_data & post_preset_data
> 
> * 리뷰 코멘트 내용이 불분명합니다.

데이터 fixture들은 모두 인라인으로 하기로 했었잖아요.. ( 지금 test 코드들은 너무 보기 힘들어요. )

hyeonjames commented on dev_up/urls.py:
```diff
@@ -33,4 +35,6 @@
     path('swagger/', schema_view.with_ui('swagger', cache_timeout=0), name='schema-swagger-ui'),
     path('redoc/', schema_view.with_ui('redoc', cache_timeout=0), name='schema-redoc'),
     path('user/', include('user.routers')),
+    path('posts', PostAPI.as_view()),
```
이 부분은 `post/routers.py` 쪽에 빈칸 형식으로 넣어주세요. 여기에 넣지 말구여

elliotyim replied:
이게 제가 원래는 post/routers.py 쪽에 넣고
path('', PostAPI.as_view())로 했더니 301 에러가 떴습니다.
이 문제를 해결해보려고 settings.py에서 APPEND_SLASH=False도 넣어보고 각 url 별로 앞에도 /를 붙여보고 떼보고 다 해봤는데도 여기서 에러나고 저기서 에러나고 해서 일단은 저렇게 해놨는데 방법을 좀 더 찾아보겠습니다.

hyeonjames replied:
`''` 대신에 `r'^$'` 넣어보세요

hyeonjames replied:
참고 : https://stackoverflow.com/questions/1316682/django-url-py-and-the-index

elliotyim replied:
음...바꾸니까 이번에는 404 에러로 바뀌네요.
urls.py의 path의 뒤쪽에 슬래시를 붙이고 안붙이고, routers.py에서 path()의 앞쪽에도 슬래시를 붙이고 안 붙이고, test에서 요청 보낼때도 슬래시를 붙여보고 안 붙여보는걸로 다 테스트 해봤는데 모든 경우에서 다 404가 뜹니다.
APPEND_SLASH도 해봤는데 효과가 없는 것 같습니다.

elliotyim replied:
어? 잠깐만...해결했습니다.
path가 아니라 re_path를 써야하는군요

hyeonjames commented on post/api/utils.py:
```diff
@@ -0,0 +1,90 @@
+from post.models import Category, Comment, Post, Tag
+
+
+class APIUtils:
```
굳이 클래스를 만들 필요가 없어보여요.
그냥 다 개별 함수로 빼주세요.
자바같은 경우야 그렇게 하겠지만, 여긴 그렇게 안해도 되요.

hyeonjames replied:
APIUtils 보다 좋은 방법이 있습니다. 따로 PR로  올려드릴게요.

hyeonjames replied:
#29 
제 PR이 머지되면 이 파일은 그냥 지우고, `common` 에 있는 함수를 이용해서 인라인으로 그냥 넣어주세요. 너무 길다 싶으면 따로 서브함수로 빼주시구요.

elliotyim replied:
넵

hyeonjames commented on post/api/utils.py:
```diff
@@ -0,0 +1,90 @@
+        return True
+
+    @staticmethod
+    def filter_by(request: Request) -> 'QuerySet[Post]':
```
`filter_by`가 여기 있을 이유가 없어요. 
왜냐하면 오직 `PostAPI`에서만 쓰이고 있고, utils의 의미가 없거든요.
오히려 코드 읽을 때 여기까지 와서 봐야 해서 복잡성만 증가합니다.

elliotyim replied:
넵

hyeonjames commented on post/api/utils.py:
```diff
@@ -0,0 +1,90 @@
+        return None
+
+    @classmethod
+    def get_list_of(cls, obj: str, *request: Request) -> 'QuerySet[Any]':
```
예전 방식이 훨씬 낫다고 봐요 저는.

hyeonjames replied:
이 함수의 필요성이 없어요. 
그나마 있다면, Post인데,  filter 함수가 따로 되어 있는 상태에서 굳이 이 함수를 쓸 이유가 없어보여요.

elliotyim replied:
filter가 model.objects.filter의 filter인지 아니면 자체적으로 옵션으로 filter하는 함수의 filter인지 잘 모르겠지만,
전자의 경우는 이 함수에 밖에 model.objects.filter를 사용하지않고 (위에는 get을 써서 모델 객체 하나만 가져옵니다.)
후자의 경우는 post 모델밖에 필터링 하지 않기에 1개 이상의 모든 모델 객체를 가져올 때 이 함수는 필요합니다.

hyeonjames replied:
> filter가 model.objects.filter의 filter인지 아니면 자체적으로 옵션으로 filter하는 함수의 filter인지 잘 모르겠지만,
> 전자의 경우는 이 함수에 밖에 model.objects.filter를 사용하지않고 (위에는 get을 써서 모델 객체 하나만 가져옵니다.)
> 후자의 경우는 post 모델밖에 필터링 하지 않기에 1개 이상의 모든 모델 객체를 가져올 때 이 함수는 필요합니다.

제가 말하는 건, Post에서는 만드셨던 `filter_by` 함수를 쓰라는 뜻이에요.
애초에 모델에 따라 다른 로직을 탄다는게 좋은 방식이 아니에요. ( 그렇게 만들기를 원하면, 추상화된 클래스를 지정하고 함수를 오버라이딩 하는 방식으로 만들어야죠. )
왜냐하면 함수 바깥의 입장에서는 제대로된 입 출력의 예상을 할 수가 없어요.

elliotyim replied:
아 그렇군요;; 확실히 함수 바깥에서 입출력의 예상이 안되는 설계는 절대 좋은 설계는 아닌 것 같네요.

hyeonjames commented on post/api/utils.py:
```diff
@@ -0,0 +1,90 @@
+        Returns an object which is active.
+        """
+        try:
+            if obj == 'Category':
```
if문으로 이렇게 분기 탈 필요가 없는 코드라고 생각해요.
모델을 바로 받아서 `model.objects.filter` 쓰면 되니까요.

elliotyim replied:
넵 근데 Tag 오브젝트의 경우 is_active가 없어서 똑같이 조건에 is_active=True를 쓰면 에러가 납니다.
그래서 if문을 주되 model이 Tag 인스턴스인 경우에만 is_active=True 조건을 주고 나머지는 전부 model.objects.filter로 하면 되겠네요.

elliotyim replied:
아 그리고 여긴 filter가 아니고 get이군요

hyeonjames replied:
> 넵 근데 Tag 오브젝트의 경우 is_active가 없어서 똑같이 조건에 is_active=True를 쓰면 에러가 납니다.
> 그래서 if문을 주되 model이 Tag 인스턴스인 경우에만 is_active=True 조건을 주고 나머지는 전부 model.objects.filter로 하면 되겠네요.

아뇨 그것도 좋은 방법이 아니에요.
그냥 차라리 매개변수를 모두  `Model.objects.get` 함수 안으로 던져서 바깥에서 is_active를 던지는게 훨씬 나아요.

elliotyim replied:
아아 그럼 함수 안에서는 전부 동일한 작동을 하고 is_active 부분만 따로 바깥에서 해주는게 맞겠군요.

hyeonjames replied:
사실 #29 에서 따로 뺄거라서 작업은 안해주셔도 됩니다.

elliotyim replied:
아 넹 봤어요
머지하시면 리베이스 할게요

hyeonjames reviewed (changes requested):
로직을 APIUtils로 넣는 건 별로 좋은 것 같진 않아요.
APIView 코드가 그냥 눈으로 봐서는 깔끔하게느껴질 수야 있겠지만, 
고작 한 두번 쓰일 코드를 다른 파일까지 와서 봐야 되서 다른 사람의 입장에서 코드를 읽는 흐름이 끊기고, 더 복잡하게 만들거든요.

그리고, 범용적인 코드를 만들꺼면, 단순히 post 안에만 동작하는게 아니라, 모든 모델에 대한 코드를 작성하는게 좋습니다.
(이에 대한건 제가 따로 작업해서 PR로 올려드릴게요. 참고해주세요.)

hyeonjames commented on tests/post/test_category.py:
```diff
@@ -0,0 +1,55 @@
+        assert response.status_code == 201
+
+    def test_create_the_same_title_category(self, api_client):
+        response1 = api_client.post('/posts/categories', data={'title': 'duplicates'})
```
사실 이 코드가 있으면 `test_create_category`가 무의미해요. (이미 체크하고 있기 때문이죠. )
하나로 합쳐주세요..

elliotyim replied:
어 뭐지? 이거 최근 푸시한 파일에는 없는 코드인데? 버그난건가? 현재 test_category에는 def test_list_categories() 밖에 없어요

elliotyim committed:
Modify `/posts` path

elliotyim committed:
Polish test codes

hyeonjames commented on tests/post/test_post.py:
```diff
@@ -0,0 +1,74 @@
+        assert response.status_code == 403
+
+    def test_delete_post(self, api_client, posts):
+        before_delete = api_client.get('/posts/1')
```
굳이 before_delete가 필요할것 같진 않네요 ( 이미 만들어 주는게 확인이 되니까요 )

elliotyim replied:
넹

hyeonjames commented on post/api/category.py:
```diff
@@ -0,0 +1,67 @@
+        Makes the category disabled.
+        The specific category ID must be required by uri resources.
+        """
+        category = APIUtils.get('Category', id=category_id)
```
카테고리 및 태그의 삭제/수정은 없는 걸로 하기로 했던걸로 기억 하는데, API가 있어야할 이유가 있나요?
아니라면 그냥 없애는게 나을 것 같습니다.
일단, 태그는 삭제하는 기능 자체가 없는게 낫고, 카테고리는 저희만 관리하는게 낫습니다.

elliotyim replied:
아 저기 위에 Outdated 태그가 붙은 애들은 옛날 파일인것 같군요. 지금은 category.py랑 tag.py에 리스팅 하는 get()만 있어요.

elliotyim committed:
Remove all http methods except for `get()` and polish codes

Change Log:
- Tag and Category API now have only get method.
- Optional parameter `url_resources` are removed.

elliotyim committed:
Remove unnecessary tests and polish codes

Change Log:
- Removed unnecessary test codes.
- format='json' lookup is added where data are required.

elliotyim committed:
Add transaction

Change Log:
- Tags now added when a post is created. (also handles transactions)

elliotyim committed:
Remove detail function from category & tag API

elliotyim committed:
Remove detail function related test codes

elliotyim committed:
Modify visibility of private functions

elliotyim committed:
Polish codes

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,111 @@
+        tags: List[Tag] = list()
+        for title in titles:
+            if Tag.objects.filter(title=title).exists():
+                tags.append(Tag.objects.get(title=title))
```
이렇게 하면 일종의 SELECT 쿼리를 두번하게 되는 거라고 볼 수 있어요. 
따라서 그냥
try ~ except 로 get을 감싸주세요.

elliotyim replied:
SELECT 두 번 하는거는 안그래도 코딩하면서 계속 걸리긴 했지요...
근데 try~ except는 무슨 말인가요?

hyeonjames replied:
`ObjectDoesNotExist` 잡아서 처리하라는 뜻이에요.
```python
try:
     tag = Tag.objects.get(title=title)
except ObjectDoesNotExist:
     tag = Tag.objects.create(title=title)
```

elliotyim replied:
아..이해했습니다. exists()랑 Tag.objects.get()이랑 두번 해서 select 쿼리를 두번한다고 하셨던거였군요?
전 태그가 2개 이상일 경우 select를 두 번하게 되니까 bulk_create()로 한번에 여러개를 만들던가 아니면 filter()로 여러개를 가져오라고 하신줄 알았네요.
이 부분은 코드를 좀 수정해서 get_or_create()를 사용해서 바꿨습니다.
(이 내용 커밋메시지에는 적는걸 깜빡해서 다른 변경사항[7f5a73e]이랑 같이 커밋되어버렸는데, 적용은 되었습니다.)

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,111 @@
+        Create a post.
+        A category ID in request data must be required.
+        """
+        key_filter = ('tags', 'tagTitles')
```
`tags`와 `tagTitles`가 굳이 둘 다 있어야할 이유가 있을까요?
전 `tags`라는 이름으로 오직 타이틀만 받을 수 있도록 하는게 좋을 것 같아요.

elliotyim replied:
넵 수정하겠습니다

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,111 @@
+        post_data: Dict[str, Any] = {k: v for k, v in request.data.items()
+                                     if k not in key_filter}
+
+        tag_titles = request.data.get('tagTitles')
```
태그는 `,`로 구분된 형태로 한번에 온다고 했지 않았나요?

elliotyim replied:
`,`요? 어떤걸 말하시는건지 잘 모르겠습니다.

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,111 @@
+        serializer = PostSerializer(post)
+        serializer.update(post, {'is_active': False})
+
+        return Response(status=status.HTTP_204_NO_CONTENT)
```
이거 그냥 return 안하면 자동으로 `NO_CONTENT` 해주지는 않나요?

elliotyim replied:
예전에 해봤는데 기본적으로 성공하면 다 200으로 보내더라구요.
그래도 혹시 세팅하는게 있을지 모르니까 찾아보겠습니다.

elliotyim replied:
계속 찾아봤는데 공식문서에서는 default 값이 200이고 따로 지정해주지 않으면 그냥 200으로 반환합니다.
return 하는 부분을 주석으로 막아봤더니 아예 500이 뜨면서 에러가 나네요

hyeonjames replied:
status코드를 직접 주는 형식은 그닥 선호하진 않는데
나중에 common에 데코레이터 같은 걸 추가하는게 좋겠네요

elliotyim replied:
> status코드를 직접 주는 형식은 그닥 선호하진 않는데
> 나중에 common에 데코레이터 같은 걸 추가하는게 좋겠네요

넵

hyeonjames commented on post/serializers.py:
```diff
@@ -0,0 +1,66 @@
+from user.models import User
+
+
+class TempUserSerializer(serializers.ModelSerializer):
```
이거 쓸 일이 없어보이는데 어디서 쓰고 있나요?

elliotyim replied:
얘가 이제 밑에 보시면 알겠지만 Serializer 중에 depth = 1로 준 애들은 단순히 필드의 id만 출력하는게 아니라 그 필드 객체의 속성들까지 추가로 출력을 해줘요. 이 때 그 필드에 해당하는 Serializer가 지정이 되어있어야 자동으로 그 필드에 대한 값을 nested 된 Serializer가 출력을 해줍니다.

ex) depth = 1을 지정 안해준 경우
{
    ....
    "category": 1,
    ....
}

ex) depth = 1 지정해준 경우
{ 
   ...
  "category": {"id": 1, "title": "test_title"},
    ...
}

우리 같은 경우에는 author 필드가 user랑 매칭되어 있는데 제 작업 브랜치에는 아직 user쪽의 Serializer가 없어서 임시로 만들어서 쓰고 있었어요. 저렇게 안하면 에러나거든요.

elliotyim committed:
Remove `before delete` on post delete test

choitaehoon1998 reviewed (changes requested):
get/post/1 같이 진행하였을때 post_id =1 에 해당하는 코멘트 내용들까지 같이 가져오는것이 좋을거같습니다 !

elliotyim commented:
> get/post/1 같이 진행하였을때 post_id =1 에 해당하는 코멘트 내용들까지 같이 가져오는것이 좋을거같습니다 !

네 수정하겠습니다.

elliotyim committed:
Add comments' data in response data of a post

Change Log:
- post response data now contain comments.
- `post` field has been removed from comment serializer as it is not necessary anymore.

elliotyim committed:
Modify request parameter name

Change Log:
- tagTitles -> tags

choitaehoon1998 commented on post/api/comment.py:
```diff
@@ -0,0 +1,66 @@
+
+
+class CommentAPI(APIView):
+    def get(self, request: Request, comment_id: int = None) -> Response:
```
get 에서 조건을 추가해줘야할것같습니다 
현재 delete 메서드에서 comment 를 지우지않고 is_active 를 false 로 변경하는데 , 
이렇게 된다면 delete 메서드를 실행한 번호에대한  get 을 실행하게되면 delete 에서 삭제되었다고 생각되는 comment 가 나오지않을까요 
사실 user api 만들다가 저도 똑같은 문제에 봉착한거같아서요

hyeonjames replied:
생각해보니 Post 에서 코멘트 가져오는 거면 Comment의 get도 쓸 일이 없을 것 같다는 생각이 드네요.

elliotyim replied:
그러네요. comment의 get도 삭제하겠습니다.

elliotyim committed:
Fix a typo

elliotyim committed:
Remove get method from comment api

hyeonjames commented on tests/post/test_post.py:
```diff
@@ -0,0 +1,117 @@
+        # in order to test transaction.
+        before_transaction = api_client.get('/posts/tags')
+        post_data['tags'] = ['tag3', 'tag4']
+        post_data['title'] = '1234567890123456789012345678901234567890123456\
```
이걸 길게 설정하신 이유가 따로 있나요? 혹시나 타이틀 길이로 오류내려고 했다면 그냥
`'1'*100` 이렇게도 가능해요

elliotyim replied:
넵 수정하겠습니다.

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,112 @@
+
+            return Response(post_data)
+        else:
+            posts = APIUtils.get_list_of('Post', request)
```
근데 이거 생각해보니 한번에 다 조회하려구요?
카테고리 같은거야 양자체가 많이 없으니 괜찮은데, Post는 다 조회하면 안됩니다
페이지 기능 넣어주세요

elliotyim replied:
작업한게 너무 많아서 페이징은 이 PR이 머지되고 나면 하려고 했는데, 생각해보니 파일이 늘어날것 같지는 않으니 그럼 여기서 작업을 해두겠습니다.

hyeonjames commented on tests/post/test_comment.py:
```diff
@@ -0,0 +1,53 @@
+
+        assert comments_length_before == 3
+        assert response.status_code == 204
+        assert comments_length_after == 2
```
이런식으로 비교하는 것보다는, `filter`를 이용해서 방금 삭제한 코멘트의 `is_active`가 `False`인지만 비교해도 충분할 것 같아요.
이 테스트는 너무 간접적인 것 같네요.

elliotyim replied:
넵

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,194 @@
+
+class TestPostFiltering:
+    pytestmark = pytest.mark.django_db
+    preset_data = {
```
이건 바깥으로 빼주시고 이런 데이터 같은 경우 대문자로 맞춰주는게 좋을 것 같네요

elliotyim replied:
넵 수정하겠습니다.

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,194 @@
+                tag_count += 1
+            elif len(tag_id) > 1 and tag_id[1] == tag['id']:
+                tag_count += 1
+        return tag_count
```
이거, tag_id가 아니라 title로 구분을 해주는게 좋을 것 같구요,
그리고 무엇보다도 이 함수 이해하는데 꽤나 오래 걸렸는데, 주석을 주시던가 아니면 좀 더 쉬운 코드로 바꾸셔야 될거 같네요.
의도는 알겠는데, 이 코드는 tag_id를 2개까지 밖에 커버할 수 없네요?, tag_id가 얼마나 오던지 커버할 수 있는 코드로 바꿔보셔요

elliotyim replied:
넵 수정하겠습니다.

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,194 @@
+        assert post1 != post2 and post1 != post3 and post2 != post3
+
+        category_id = categories[0].id
+        response = api_client.get(f'/posts?category={category_id}')
```
카테고리, 태그는 이미 타이틀 자체가 identity를 보장합니다.
따라서 id는 최대한 쓰지말고 타이틀로만 필터링 해주세요.

elliotyim replied:
넵

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,194 @@
+
+        post1 = api_client.post('/posts', post1_data, format='json').data
+        post2 = api_client.post('/posts', post2_data, format='json').data
+        post3 = api_client.post('/posts', post3_data, format='json').data
```
너무 중복된 코드가 많습니다 특히 이 업데이트 하는 부분이요.
이 부분은 그냥 같은 파일의 fixture로 만들어주는게 좋을 것 같아요.

elliotyim replied:
제가 아무리 생각해도 이 부분을 잘못 이해한것 같아서 그러는데 어떤 방식으로 해야하는지 다시 한번만 알려주심 안될까요?
conftest.py에서 한 것 처럼 fixture 데코레이터를 붙이고 주입시키는 방식으로 하는건가요?

elliotyim committed:
Add a fixture of 100 of posts for testing pagination

elliotyim committed:
Add `is_active` field into comment serializer for testing delete method

elliotyim committed:
Remove APIUtils module from api

elliotyim committed:
Modify comment test case

Change Log:
- Test of comparison comments length is changed to testing if it is active or not

elliotyim committed:
Apply common API

Change Log:
- APIUtils are not used anymore.
- If `content` parameter is empty, put method of comment api will cause an error.

elliotyim committed:
Modify post test codes and polish codes

Change Log:
- Long data in `title` are fixed properly.
- Pagination test codes are added.
- If there is no parameter data in patch method, post api will cause an error.
- Only when filtering, values of `category` & `tags` parameters should be `title`
  from now on.

elliotyim committed:
Implement pagination and polish codes

Change Log:
- APIUtils are replaced by common API.
- Pagination is supported.

elliotyim committed:
Polish codes

elliotyim committed:
Separate API and rename serializers make them ready for the swagger

Change Log:
- Separated PostAPI & CommentAPI and added mixins in order to `drf_yasg` generates
  paths properly.
- Rename ListSerializers: `drf_yasg` generates model names with their serializer's name
  ex) PostListSerializer -> model: PostList,
- As the first line of docstring is the description title of API's, a blank line is added to
  separate title from content.

elliotyim committed:
Fix response status code properly

After patch or put, status code should be 200.
Athough any tags could be created as the result of patch of `post`,
it is not be considered at the moment. (In this case, status code should be 201)

elliotyim committed:
Fix docstring of post API properly

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,171 @@
+        post1_data['category'] = categories[0].id
+        post2_data['category'] = categories[1].id
+        post3_data['category'] = categories[0].id
+        self.set_posts_with(api_client, post1_data, post2_data, post3_data)
```
테스트에서 원하는건 각각 category, tag, date 등등 필터링이 잘 되는지 확인하려고 하는 거잖아요?
그럼 굳이 테스트 마다 post를 업데이트 시켜줄 필요가 없지요.
그냥  모든 컬럼을 한꺼번에 바꿔서 update하는 fixture 하나만 두면 훨씬 더 짧아지죠

elliotyim committed:
Polish codes

elliotyim committed:
Refactor post filtering test codes

Removed duplicates by making preset post fixtures.

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,137 @@
+        tags = [tag for tag in result['tags'] if tag['title'] in title_filters]
+        return len(tags)
+
+    def test_post_on_category_filtering(self, api_client, users, posts, categories, tags):
```
`users`, `tags`는 필요가 없어요

elliotyim replied:
넵 수정하겠습니다.

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,137 @@
+        assert first_result['category']['title'] == category_title
+        assert second_result['category']['title'] == category_title
+
+    def test_post_on_tag_filtering(self, api_client, users, posts, categories, tags):
```
`users`,`categories`는 필요가 없어요

elliotyim replied:
넵 수정하겠습니다2

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,137 @@
+        response = api_client.get(f'/posts?tags={tag_title1}&tags={tag_title2}')
+        assert response.status_code == 400
+
+    def test_post_on_category_tag_filtering(self, api_client, users, posts, categories, tags):
```
여기도 마찬가지로 필요없는 것들 다 빼주세요

elliotyim replied:
넵 수정하겠습니다3

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,137 @@
+        assert response.status_code == 200
+        assert len(response.data) == 2
+
+        first_result, second_result = response.data[0], response.data[1]
```
곡 첫번째, 두번째만 검사할 필요가 있나요? 그냥 `response.data`에 있는 거 다 검사하면 되지 않을까 싶어요.

elliotyim replied:
넵 수정하겠습니다4

elliotyim committed:
Polish codes

Change Log:
- Removed redundancy
- Checking all IDs and category titles

elliotyim committed:
Polish codes

elliotyim committed:
Fix a mypy error

elliotyim committed:
Polish test codes

Change Log:
- Renamed data parameter `perPage` to pageSize.
- Made test case more clearly.

elliotyim committed:
Use specific serializer

elliotyim committed:
Polish post api codes

Change Log:
- Parameter `perPage` -> `pageSize
- Use specific serializer
- When retrieving a post, the post api is querying twice before.
   (one for Post, another one for Comment)
   This problem is solved by using a nested serializer.
- Use `related_name` lookup in comment model field to make PostCommentSerializer
   recognise nested field in Post model.

elliotyim committed:
Polish test codes

elliotyim committed:
Modify APIView to viewsets

elliotyim committed:
Polish codes

elliotyim committed:
Polish codes

elliotyim committed:
Fix a typo

elliotyim committed:
Optimise serializers

- Removed both PostCreateSerializer & CommentCreateSerializer. All the fields in
   each serializer are managed depending on the situation.

elliotyim committed:
Polish codes

elliotyim commented on post/models.py:
```diff
@@ -46,13 +44,12 @@ class Comment(models.Model):
-    parent_comment = models.ForeignKey('self', on_delete=models.PROTECT, null=True)
+    post = models.ForeignKey('Post', on_delete=models.CASCADE, related_name='comments')
+    parent_comment = models.ForeignKey('self', on_delete=models.PROTECT, null=True, default=None)
     author = models.ForeignKey(User, on_delete=models.SET_NULL, null=True)
```
코멘트 작성 시 post 값을 안 줘도 코멘트가 작성되는 버그가 있기에 null=False로 바꾸고
post가 지워질 일은 없지만, 지워진다면 코멘트도 쓸모 없어지므로 CASCADE를 적용했습니다.
author의 경우는 jwt를 적용하게 되면 null을 넣는 상황이 아예 없어지므로 그대로 두었습니다.

hyeonjames replied:
Post가 없어질 일도 없으니 CASCADE로 가도 상관은 없는데요, 이걸 해결 방법으로 내놓은 거라면 좀 이상하군요
그런 버그가 있다면 Post 모델을 바꾸는게 아니라 API쪽에서 체크를 해줄 수 있고 blank라는 옵션으로 해결하는 방법도 있구요.

지금 와서 생각해보니 SET_NULL보다는 CASCADE가 가져오는 장점이 더 많을 것 같아서 이대로 가는 걸로 하셔도 되지만, 이건 저번에도 말씀 드렸던건데 모델의 옵션을 바꾸는 건 단순히 코드 몇 줄 고치는 거랑 무게가 많이 다르니까, 다음에는 다른 방법을 먼저 찾아보셨으면 좋겠네요.

elliotyim replied:
> Post가 없어질 일도 없으니 CASCADE로 가도 상관은 없는데요, 이걸 해결 방법으로 내놓은 거라면 좀 이상하군요
> 그런 버그가 있다면 Post 모델을 바꾸는게 아니라 API쪽에서 체크를 해줄 수 있고 blank라는 옵션으로 해결하는 방법도 있구요.
> 
> 지금 와서 생각해보니 SET_NULL보다는 CASCADE가 가져오는 장점이 더 많을 것 같아서 이대로 가는 걸로 하셔도 되지만, 이건 저번에도 말씀 드렸던건데 모델의 옵션을 바꾸는 건 단순히 코드 몇 줄 고치는 거랑 무게가 많이 다르니까, 다음에는 다른 방법을 먼저 찾아보셨으면 좋겠네요.

아...모델은 함부로 바꾸는게 아니군요. 주의하겠습니다;;

hyeonjames commented on post/serializers.py:
```diff
@@ -0,0 +1,83 @@
+    createdDate = serializers.DateTimeField(source='created_date')
+    comments = CommentSerializer(many=True)
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
```
이 기능은 중복이 되잖아요? 그럼 사실 common에 이런 기능을 하는 serializer를 만들어 주어 상속시키면 이런 중복을 피할 수 있어요.

elliotyim replied:
오 그렇군요! 생성자도 상속이 되는줄 몰랐습니다. 그렇게 바꿀게요.

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,153 @@
+
+        if 'page' in params:
+            no = int(params['page'])
+            page_size = params['pageSize'] if 'pageSize' in params else 20
```
params.get('pageSize', 20) 을 하면 되지여

elliotyim replied:
천재다...그렇게 바꾸겠습니다.

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,153 @@
+        posts = posts.filter(is_active=True)
+        serializer = PostSerializer(posts, many=True, fields=allowed_fields)
+
+        if 'page' in params:
```
앗, 페이징은 옵셔널로 적용하면 안되요. 꼭 적용해야 되요
페이지 사이즈 또한 최대값을 설정해주세요.
자칫하면 트래픽 공격을 위한 API가 될 수 있어요.

elliotyim replied:
~~페이지 사이즈 최대값이라는게 2147483647 이런거 막으라는 얘긴가요?~~
네

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,153 @@
+from post.models import Post, Tag
+from post.serializers import PostSerializer
+
+POST_FIELDS = set([
```
{}를 이용해 set을 선언해주세요

elliotyim replied:
넵

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,153 @@
+        tag_titles = request.data.get('tags')
+        if tag_titles is not None:
+            tags = self.__create_tags_with(tag_titles)
+            request.data.update(tags=[tag.id for tag in tags])
```
request.data를 업데이트를 시켜주는 건 별로 좋은 생각이 아니에요. 
따로 dictionary를 만들어서 처리해주세요.

hyeonjames replied:
request 자체를 immutable 한 객체라 생각하고 사용해주세요, serializer 쪽에도 request.data를 넘기지 말고 따로 dictionary를 만들어 넣어주셔요.

elliotyim replied:
넵 둘 다 적용하겠습니다.

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,132 @@
+            number_of_tags = self.__count_tags_of(result, *tag_titles)
+            assert number_of_tags == len(tag_titles)
+
+    def __check_data_duplication_of(self, response):
```
사실 test안에 있는 것들은 private 처리가 크게 필요하진 않은 것 같아요.
아니면 그냥 클래스 밖으로 함수를 빼는 것도 하나의 방법인 것 같네요.

elliotyim replied:
그냥 클래스 밖으로 빼도록 하겠습니다.
자바 때 하던 습관이 남아있어서 클래스에 넣는게 깔끔해보였나봅니다 ㄷㄷ;;

hyeonjames reviewed (changes requested):
거의 끝나고 있는 것 같네요, 힘내셔요

elliotyim committed:
Modify test codes

Change Log:
- Query parameter `page` is now mandatory.
  Hence, testing `get list of posts without page no.` is removed
- All the util functions such as `count tags of` are moved outside of Test class and
  modified visibility.

elliotyim committed:
Implement SerializerFieldAdjuster

It will be the super class for serializers which need to remove a few of their fields.

elliotyim committed:
Make another dictionary to pass request data

elliotyim committed:
Polish codes

Change Log:
- Change the way of declaration of POST_FIELDS (set() -> {})
- Make it simple that getting page size
- Make the page size has a maximum size: 50.
- Make another dictionary to pass or update request data.
- If page or page size is not integer, an error will be raised.
- If page is not in parameter, an error will be raised.

elliotyim commented:
common에 따로 빼놓은 SerializerFieldAdjuster 혹시 더 좋은 이름 있으면 추천받습니다.
지금 이름은 뭔가 2% 부족한 것 같아요.

hyeonjames commented on common/serializers.py:
```diff
@@ -0,0 +1,19 @@
+from rest_framework import serializers
+
+
+class SerializerFieldAdjuster(serializers.ModelSerializer):
```
Adjuster라는 단어가 굉장히 낯설어요, 있는 단어 같기는 하지만 주로 쓰이는 단어가 아니라 좋은 네이밍은 아니지 싶습니다
`FilteredSerializer` 같은게 어떤가 싶네요,
어짜피 같은 Serializer 니까 말이죠

elliotyim replied:
넵 수정하겠습니다.

hyeonjames commented on common/serializers.py:
```diff
@@ -0,0 +1,19 @@
+        fields = kwargs.pop('fields', None)
+
+        # Instantiate the superclass normally
+        super(SerializerFieldAdjuster, self).__init__(*args, **kwargs)
```
그냥 `super()` 쓰셔도 됩니다

elliotyim replied:
어 그러네요? ㅋㅋ 근데 왜 drf 예시에는 이렇게 써져있었지?;;

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,163 @@
+            tags = self.__create_tags_with(tag_titles)
+            post_data.update(tags=[tag.id for tag in tags])
+
+        allowed_fields = POST_FIELDS - {'createdDate', 'comments'}
```
저번에도 말씀 드렸는데, 
허용되지 않은 필드를 정하는 것보다는 허용하는 필드를 정하는게 낫다고 했지요. 
이름은 allowed_fields지만 사용하는 방식은 지금 허용되지 않는 필드를 지정해준 거잖아요?

hyeonjames replied:
이 방법은 우리가 자칫 실수하면 의도치 않은 컬럼을 수정하는 권한을 사용자에게 주게 되는 문제가 있어요

elliotyim replied:
어 뭐지? 진짜 그러네요. 수정하겠습니다.

elliotyim committed:
Polish codes

elliotyim committed:
Make the `allowed fields` indicate fields directly

elliotyim committed:
Remove the unnecessary constant

elliotyim committed:
Bring `allowed fields` out from API classes

This process will also reduces redundancy when generating swagger with decorators.

elliotyim committed:
Fix the exception error properly

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,184 @@
+}
+
+CREATE_FIELDS = {
+    'id',
```
id는 포함시킬 필요가 없지요..

elliotyim replied:
현재 API에서 Response 객체를 보낼 때 지정한 필드들만 값을 표시해줍니다.
그래서 CreateSerializer라고 해도 'id' 필드를 지정해줘야지 아래와 같이 방금 생성한 레코드의 PK까지 포함해서 표시해줍니다. (만약 지정하지 않으면 표시 안해줌)
```
{
    "id": 6,
    "title": "글쓰기 테스트 한번 더",
    "content": "내용!",
    "location": "장소",
    "capacity": 5,
    "date": "2020-04-13",
    "timeOfDay": 1,
    "author": 1,
    "category": 1,
    "tags": [
        7
    ]
}
```

어차피 ID나 PK는 serializer 단에서 체크해서 막아주기 때문에 입력값으로 넣어도 수정 못해서 괜찮을듯 합니다.

근데 혹시 제가 rebase하고 강제 푸시하면 커밋이 어긋나는건가요? 저번에도 그랬는데 이번에도 Outdated된 파일이네요. 이 부분 이미 serializer를 나누는 것으로 해결이 됐는데...

hyeonjames replied:
예전에 코멘트 달았다가 완료를 안했던거 같네여

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,184 @@
+            serializer = PostSerializer(paginated_posts, many=True, fields=LIST_FIELDS)
+
+        else:
+            raise ValidationError(detail='Page parameter must be required.')
```
page 파라미터는 require이 아니라 default로 주는게 좋을 것 같아요.

elliotyim replied:
넵 수정하겠습니다. default는 page=1로 할께요

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,184 @@
+        will be returned.
+        """
+        params = request.GET
+        posts = filter_exists(Post, params.dict(),
```
한줄에 한 파라미터씩 넣어주세요

elliotyim replied:
```
posts = filter_exists(Post, params.dict(),
                              date__gte='startDate',
                              date__lte='endDate',
                              time_of_day='timeOfDay',
                              location__contains='location',
                              category__title='category',
                              ).order_by('id')
```
이걸 아래처럼 고치라는 말인가요?

```
        posts = filter_exists(Post, params.dict(), date__gte='startDate')
        posts = filter_exists(posts, params.dict(), date__lte='endDate')
        posts = filter_exists(posts, params.dict(), time_of_day='timeOfDay')
        posts = filter_exists(posts, params.dict(), location__contains='location')
        posts = filter_exists(posts, params.dict(), category__title='category')
        posts = posts.order_by('id')
```

hyeonjames replied:
그럼 filter_exists를 만든 이유가 없지요
`params.dict(),` 앞에 한칸 띄워 달라는 뜻이에요
그리고 줄 정렬 맞춰주세요. autopep8가 맞춰줄텐데

hyeonjames replied:
기본적으로 너무 길어서 한 줄 이상으로 가야 할 경우에는 한 라인당 한개씩의 인자만 넣는게 좋습니다.
그리고 인자의 시작 부분도 맞춰주셔야 되구여
```python
        cls = type(f'{origin.__module__}.{origin.__name__}',
                   tuple(inherited),
                   dict(Meta=meta_class))
```

elliotyim replied:
> 그럼 filter_exists를 만든 이유가 없지요
> `params.dict(),` 앞에 한칸 띄워 달라는 뜻이에요
> 그리고 줄 정렬 맞춰주세요. autopep8가 맞춰줄텐데

저거 복붙 하면서 시작 줄이 어긋난것 같네요. 원래는 밑의 예시처럼 시작부분이 맞춰져 있어요.
암튼 밑에것처럼 수정하겠습니다.

hyeonjames replied:
아 그렇네여

elliotyim committed:
Change back all the serializers that inherited `FilteredSerializer`

elliotyim committed:
Split each serializers

elliotyim committed:
Apply each serializers properly

elliotyim committed:
Fix a test error

elliotyim committed:
Fix loop bug

After deletion of a comment, API should not return any value,
otherwise it will loop.

elliotyim committed:
Alter the way of deletion test of comment

hyeonjames commented on post/serializers/category_serializers.py:
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
```
성빈님 네이밍 지을때 신경을 좀 써주세요. 계속 중복된 이름을 많이 넣으시는 것 같네요.
파일 이름 뒤에 _serializers는 필요가 없잖아요 ( 폴더가 이미 대신 해주고 있는데 )

elliotyim replied:
네 알겠습니다.

elliotyim committed:
Rename serializers

Change Log:
- Rename *_serializers -> *
- Allocate one line for each field.
- Remove unnecessary codes.

elliotyim committed:
Polish codes

Change Log:
- Allocate one line for each parameter.
- Add `author` into parameter exception rule for patching.
  For some reason, `serializer.update()` works different way compare to
  `serializer.save()`, all the parameters need to be checked before update.

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,147 @@
+        try:
+            posts = posts.filter(is_active=True)
+            no = int(params.get('page', 1))
+            page_size = params.get('pageSize', 20)
```
`page_size = max(params.get('pageSize',20),50)` 하면 되지여

elliotyim replied:
와...진짜 코드 간결화의 극한을 보여주시는군요. 생각도 못했네..
난 언제쯤 저렇게 짤 수 있게 될까 ㄷㄷ;;
수정하겠습니다.
~~아 그리고 max가 아니라 min이 오타난것 같습니다.~~

hyeonjames replied:
아 제가 잘못 봤네요 저렇게 하면 무조건 50페이지가 될거 같은데
`min(int(params.get('pageSize',20)), 50)` 입니다

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,147 @@
+        """
+        if not request.data:
+            raise ParseError(detail='At least one field must be required to update the post.')
+        elif request.data.get('id') or request.data.get('pk'):
```
`'id' in request.data or 'pk' in request.data` 가 조금 더 직관적이고 좋을 것 같아요

elliotyim replied:
넵 수정하겠습니다.

hyeonjames commented on post/serializers/post.py:
```diff
@@ -0,0 +1,95 @@
+    class Meta:
+        model = Post
+        depth = 1
+        fields = ('id',
```
여긴 한줄 한개 안해도 되여.. 그냥 한줄에 여러개 넣는게 보기 편할거 같아요
함수 호출 같은 경우 너무 길어지는 경우 한줄에 한 인자씩 넣지만, 이런 단순 데이터는 줄로 바꾸지 않아도 됩니다.

elliotyim replied:
넵
그럼 두 줄 정도로 기존에 있었던 느낌으로 바꿀까요?

hyeonjames replied:
넵 그렇게 해주세요

hyeonjames commented on post/serializers/__init__.py:
```diff
@@ -0,0 +1,18 @@
+
+from user.models import User
+
+from .category import *  # noqa 401
```
relative 임포트는 안하는게 좋습니다.
`post.serializers.category` 형식으로 선언해주세요.

elliotyim replied:
아하 그렇군요.
아 그리고 폴더의 __init__.py로 선언할 때 import 하는거 말고 다른 좋은 방법이 혹시 있나요?
noqa 401을 붙이는게 너무 불편해서...
다른 프로젝트에서 찾아서 발견한 방법이 이 방법이긴 한데, 거긴 mypy를 안 써서 그냥 넘어갔는데 혹시 mypy를 쓸 때 noqa 401을 붙이지 않고도 깔끔하게 사용할 수 있는 다른 방법이 있는지 궁금합니다.

hyeonjames replied:
아뇨 딱히 그런 방법은 없는 것 같아요
어쩔 수 없는 예외케이스 일거에요.

elliotyim committed:
Change relative paths into absolute paths

elliotyim committed:
Modify allocated lines of each field of serializers

elliotyim committed:
Polish codes

elliotyim commented:
serializer의 필드를 나열할 때, `두 줄 정도`의 기준이 좀 애매하여
한 라인 당 4개의 필드까지 나열하는 것으로 고쳤습니다.

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,146 @@
+
+        A specific post ID must be required by uri resources.
+        """
+        if not request.data:
```
사소한 점인데 뒤의 `patch_data` 를 앞으로 옮기고 `request.data` 대신에 `patch_data` 를 써야 일관성이 좀 맞을 것 같네요

elliotyim replied:
> 사소한 점인데 뒤의 `patch_data` 를 앞으로 옮기고 `request.data` 대신에 `patch_data` 를 써야 일관성이 좀 맞을 것 같네요

그쵸 그게 맞죠
수정하겠습니다

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,146 @@
+        the tag objects will be returned.
+        """
+        tags: List[Tag] = list()
+        for title in titles:
```
trim() 과 빈 문자열 무시하는 로직이 있으면 좋겠네요.

elliotyim replied:
> trim() 과 빈 문자열 무시하는 로직이 있으면 좋겠네요.

아...이게 제가 파라미터에서 tag title을 뽑아오니까 trim이 적용이 안됐군요.
기존에 PK로 받아올 때는 자동으로 trim을 적용하던데..
위의 두 가지 사항 수정하겠습니다.

hyeonjames commented on post/routers.py:
```diff
@@ -0,0 +1,43 @@
+urlpatterns = [
+    path('/categories', CategoryAPI.as_view({'get': 'list'})),
+    path('/tags', TagAPI.as_view({'get': 'list'})),
+    path('/comments/<int:comment_id>', CommentAPI.as_view({
```
여기서 dictionary를 주지 않고 구현된 함수 위에서 `@method('GET')` 이나 `@GET` 같은 데코레이터로 꾸민 뒤 여기에는 간단하게 `~~API.as_view_with_method()` 같은 걸로 알아서 처리해 줄수 있도록 하면 좋을것 같네요.
이 PR 머지 된 후에 후속작업으로 한번 진행 해보실래요?

elliotyim replied:
> 여기서 dictionary를 주지 않고 구현된 함수 위에서 `@method('GET')` 이나 `@GET` 같은 데코레이터로 꾸민 뒤 여기에는 간단하게 `~~API.as_view_with_method()` 같은 걸로 알아서 처리해 줄수 있도록 하면 좋을것 같네요.
> 이 PR 머지 된 후에 후속작업으로 한번 진행 해보실래요?

넵
안그래도 데코레이터에 익숙해져야 하니까 연습도 해볼겸
PR 하나 더 만들어서 작업해보겠습니다.

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,135 @@
+        assert len(response.data) == 2
+
+        for result in response.data:
+            assert start_date <= result['date'] and result['date'] <= end_date
```
이거 파이썬에서는 `start_date <= result['date'] <= end_date` 가 가능할꺼에요

elliotyim replied:
> 이거 파이썬에서는 `start_date <= result['date'] <= end_date` 가 가능할꺼에요

역시 갓 파이썬...진짜 합리적이고 좋네요
수정하겠습니다.

hyeonjames commented on tests/post/test_post_filtering.py:
```diff
@@ -0,0 +1,135 @@
+def check_data_duplication_of(response):
+    ids = set()
+    for result in response.data:
+        ids.add(result['id'])
```
사소하지만 이거 `ids = {result['id'] for result in response.data}` 한줄로 가능해여

elliotyim replied:
> 사소하지만 이거 `ids = {result['id'] for result in response.data}` 한줄로 가능해여

사소하지 않아요 ㅋㅋ
이런 디테일을 대체 어디가서 배울 수 있단 말이죠?
이런게 진짜 꿀팁이지 ㅋㅋ
바로 수정하겠습니다.

hyeonjames reviewed (changes requested):
거의 다 온거 같아요. 이제 거의 자잘한 수정사항 밖에 없는 것 같네요 :)
조금 만 더 힘내시길  👍

elliotyim committed:
Modify post API

Change Log:
- If blank is given by request parameters, it will be ignored.
- Refactoring

elliotyim committed:
Polish codes

elliotyim committed:
Modify methods: `POST`, `PUT`, `PATCH`

Change Log:
- After creating or updating posts or comments, the API returns detail messages
  instead of its own model data.

elliotyim committed:
Polish codes and fix bugs

Change Log:
- Applied ValidationError properly.
- Removed redundant codes.
- Tag searching is now working properly with subset words.

elliotyim committed:
Polish codes

elliotyim committed:
Removed `post` field in `CommentSerializer`

It is redundant.

elliotyim committed:
Polish codes

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,144 @@
+
+        return posts
+
+    def __convert(self, tag_titles: List[str] = None) -> Optional[List[int]]:
```
List[str]은 None 이 되면 안되요. Optional[List[str]]로 선언해주세요

hyeonjames replied:
그냥 `__create_tags_with` 랑 합쳐주세요. 여기 말고는 __create_tags_with 이 안쓰이네요

elliotyim replied:
> 그냥 `__create_tags_with` 랑 합쳐주세요. 여기 말고는 __create_tags_with 이 안쓰이네요

넵 수정하겠습니다.

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,144 @@
+
+        return posts
+
+    def __convert(self, tag_titles: List[str] = None) -> Optional[List[int]]:
```
앞에 언더바는 하나만 넣는게 어때요?

elliotyim replied:
> 앞에 언더바는 하나만 넣는게 어때요?

넹. 근데 보통 언더바 하나가 표준인가요?
제가 알고 있는 거로는 `_`가 Java로 따지면 Protected고 `__`가 private이라서 두 개 쓰는 건줄 알았는데, 사실 잘 몰라서...

hyeonjames commented on post/api/post.py:
```diff
@@ -0,0 +1,144 @@
+        will be returned.
+        """
+        params = request.GET
+        posts = filter_exists(Post,
```
filter_exists는 queryset도 넣을 수 있어요. 따라서
`Post.objects.filter(is_active=True)` 를 여기에 넣는게 어때요?

elliotyim replied:
> filter_exists는 queryset도 넣을 수 있어요. 따라서
> `Post.objects.filter(is_active=True)` 를 여기에 넣는게 어때요?

넵 수정하겠습니다.

hyeonjames reviewed (approved):
일단 코멘트 드린거 수정만 해주시고, 리베이스 후 머지하는 걸로 하죠, 추후에 수정사항은 따로 PR로 요청해주세요.
거의 한달 반동안 수고 많으셨습니다 ㄷㄷ

elliotyim committed:
Polish codes

Change Log:
- Parameter type of tag_titles: List[str] in _convert() is changed into Optional.
- _created_tags_with() is merged into _convert()
- filter_exists() has active posts instead of all the posts.

elliotyim committed:
Set `APPEND_SLASH=False`
